tactics/tactics.ml

```diff
@@ -1606,7 +1606,9 @@ let descend_in_conjunctions avoid tac (err, info) c =
                   (assert_before_gen false (NamingAvoid avoid) pt)
                   (fun _ -> Proofview.tclZERO ~info err))
                 [Proofview.tclORELSE
-                   (refiner ~check:true EConstr.Unsafe.(to_constr p))
+                  (Proofview.Goal.enter begin fun gl ->
+                    convert_leq pt (Proofview.Goal.concl gl) <*> Refine.refine ~typecheck:false (fun h -> (h, p))
```

Comment:
Do we need the convert here? `assert` was also called with `pt` so it would be strange if there was a difference IIUC

Reply:
I think it is useless, but it doesn't hurt to check twice. The code above is a slightly unsafe variant of `exact`.

Reply:
It makes the code mildly more complicated especially since it's also the only use for the new `enter`

Reply:
To repeat myself, methinks that `apply` should be rewritten mostly from scratch, which would suppress a lot of code of this kind. This is not the only instance where we do *ceinture et bretelles*.

Reply:
To avoid confusing the reader we should not add useless code imo

Reply:
Let's remove the check if you insist though, it shouldn't hurt. If the CI is green I believe this should dissipate any irrational fear.